Encourage SASL over NickServ when available, tidy up

digitalcircuit opened the pull request:
## In short
* Suggest SASL in place of NickServ
  * Place ```Use SASL Authentication``` above ```Auto Identify```
  * Mark SASL as ```Supported by network``` when connected and server supports it
  * Add ```Details...``` button to explain why SASL is or is not supported
  * Add explanations in tooltips
  * All combinations still allowed, just a visual design change
* Tidy up capability handling code
* Fix SASL appearing active when ```Use SASL Authentication``` was unchecked

Criteria | Rank | Reason
---------|---------|-------------
Impact | ★★☆ *2/3* | User-facing, clarifies, answers "How do I identify before joining?"
Risk | ★★☆ *2/3* | Misleading wording if server capabilities aren't detected correctly
Intrusiveness | ★☆☆ *1/3* | Small set of changes, shouldn't interfere with other pull requests

## Rationale
In most cases, SASL authentication offers a better experience than NickServ identification.  As authentication happens before joining channels, channels requiring registration won't block attempts and hostmask cloaking will apply to ```JOIN``` messages.  Unfortunately, not all servers support SASL.

Quassel should make use of the IRCv3 capability list sent from the core to client in [pull request #198](https://github.com/quassel/quassel/pull/198 ) to check if SASL is available or not, then update the settings interface accordingly.

As networks may be disconnected or not accurately specify support for SASL (*e.g. temporary unavailability*), Quassel should not force configuring SASL one way or the other.

## Implementation
* Suggest SASL in place of NickServ
  * Rearrange the Network ```Auto Identify``` tab to place ```Use SASL Authentication``` at the top.  People tend to prioritize items shown first.
  * Add new row with icon and label saying ```Supported by network```, ```Not currently supported by network``` (*uses warning icon*), ```Cannot check if supported when disconnected```, or ```Could not check if supported by network```
  * Add new button ```Details...``` to explain each of the above states, including suggestions
  * Add tooltips to authentication options, including recommending SASL if you need to identify before joining channels (*this gets asked a lot in ```#quassel```*).
  * All past combinations of SASL and NickServ-based authentication are still allowed
  * *Also handles capability updates while the settings dialog is open!  That was.. interesting to figure out.*
* Tidy up capability handling code
  * Move ```IrcCap::SaslMechs::maybeSupported``` to ```Network::saslMaybeSupports```, simplifying usage
  * Add ```capAvailable``` to ```Network```, similar to ```capEnabled```, useful for showing elsewhere whether or not a certain capability could've been requested
  * Hide SASL EXTERNAL notice while the client's still loading
* Fix SASL appearing active when ```Use SASL Authentication``` was unchecked
  * Avoid conflict between ```sslUpdated()``` and ```sasl``` QGroupBox state
  * Simplify widget enabling/disabling by using an invisible frame

Explanations shown via ```Details...``` button:
* Messages
  1.  **SASL supported by network**:  The network "{network name}" supports SASL.  In most cases, you should use SASL instead of NickServ identification.
  2.  **SASL not currently supported by network**:  The network "{network name}" does not currently support SASL.  However, support might be added later on.
  3.  **Cannot check if SASL supported when disconnected**:  Quassel cannot check if "{network name}" supports SASL when disconnected.  Connect to the network, or try using SASL anyways.
  4.  **Could not check if SASL supported by network**:  Quassel could not check if "{network name}" supports SASL.  This may be due to unsaved changes or an older Quassel core.  You can still try using SASL.
* Bottom text, same for all
  * *SASL is a standardized way to log in and identify yourself to IRC servers.*

## Examples
### After - SASL placed above, encouraged
![Network settings dialog on client showing the Auto Identify tab.  'Use SASL Authentication' is placed above 'Auto Identify', icon and label indicate it's preferred](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20settings%20v2%20-%20new.png#v1 )

### Before - SASL placed below as if secondary
![Network settings dialog on client showing the Auto Identify tab.  'Auto Identify' is placed above 'Use SASL Authentication', no icon and label to indicate it's preferred](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20settings%20v2%20-%20old.png#v1 )

### SASL messages when supported
*Server announces support for SASL via ```CAP LS```, and if specific support is advertised, ```SASL PLAIN``` is supported*
![Network settings dialog on client showing the Auto Identify tab.  At the bottom of the 'Use SASL Authentication' box, an information icon procedes the message 'Supported by network'.  The 'Details...' button shows message #1 from Implemention.](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20SASL%20v2%20-%20connected%2C%20advertised.png#v4 )

### SASL messages when unsupported
*Server did not announce support for SASL via ```CAP LS```, or if specific support is advertised, ```SASL PLAIN``` was not supported*
![Network settings dialog on client showing the Auto Identify tab.  At the bottom of the 'Use SASL Authentication' box, a warning icon procedes the message 'Not currently supported by network'.  The 'Details...' button shows message #2 from Implemention.](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20SASL%20v2%20-%20connected%2C%20not%20advertised.png#v4 )

### SASL messages when disconnected
*Not connected to server*
![Network settings dialog on client showing the Auto Identify tab.  At the bottom of the 'Use SASL Authentication' box, an information icon procedes the message 'Cannot check if supported when disconnected'.  The 'Details...' button shows message #3 from Implemention.](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20SASL%20v2%20-%20disconnected.png#v4 )

### SASL messages when unknown
*Network was added without being saved, core does not support capability negotiation, etc*
![Network settings dialog on client showing the Auto Identify tab.  At the bottom of the 'Use SASL Authentication' box, an information icon procedes the message 'Could not check if supported by network'.  The 'Details...' button shows message #4 from Implemention.](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20SASL%20v2%20-%20unknown%2C%20changes%20not%20saved.png#v4 )

### All Auto Identify tooltips
* **Use SASL Authentication:** Authenticate using your nickname and password before joining any channels
  * **Account:** Account name, often the same as your nickname
  * **Password:** Account password
* **Auto Identify:** Authenticate to services using your password.  Use SASL instead to identify before joining channels.
  * **Service:** Service user to send your password to, usually NickServ
  * **Password:** Account password

![Network settings dialog on client showing all tooltips in the Auto Identify tab.](https://zorro.casa/go/db/Hosting/Utilities/Quassel/Development/pr/ft-suggest-sasl/Network%20settings%20-%20tooltips.png#v4 )

CarrotEngineer commented:
The documentation for this PR is great as per usual!

I'm a bit unsure about the "might not work" language. In the context of the PR / people knowledgeable about the protocol you probably know why, but I can imagine being frustrated as a general user, about not knowing why it doesn't work.

It's effectively showing whether the server advertises SASL or not, so I suspect having "preferred" show up when the server has actively advertised SASL would be enough - no "might not work" case at all.

digitalcircuit commented:
After discussion in ```#quassel```, rather than try to explain the nuances of SASL availability in the tiny space of the ```QGroupBox```, it might be better to just mark it as ```... (supported by server)``` when available and keep the default text otherwise.  Placing SASL above Auto Identify might do enough to imply it's preferred.

Instead, the [default colors pull request](https://github.com/quassel/quassel/pull/266 ) should change the ```Error message``` color to something different than the ```Server message``` color, so the [```SASL authentication currently not supported by server``` message](https://github.com/quassel/quassel/blob/b5b598dd7fbd64a72ffdeaa6df39a97100dd2810/src/core/corenetwork.cpp#L1221-L1223 ) stands out more.

AlD commented:
Great PR as always.

Please don't modify the title in the settings widget, though. I don't think people usually expect this to be dynamic content and may find it confusing.
I'd suggest to add ℹ (INFORMATION SOURCE, U+2139)/⚠ (WARNING SIGN, U+26A0) or something more Qt-appropriate with a dynamic tooltip. Even a plain old button labelled "check current server" creating a dialog would work imho.

The tooltip/dialog should include texts like

> ℹ SASL authentication is supported by the IRC server you are currently connected to (irc.server.name).

> ℹ Quassel has not been able to detect if this network supports SASL authentication. This may be because Quassel is not currently connected to it.

> ⚠ The IRC server you are currently connected to (irc.server.name) does not seem to support SASL authentication.

digitalcircuit commented:
Agreed!  I'll try to look for any Qt UI patterns to mimic and experiment with this in the next few days.

I'd lean towards a solution that's visible and simple by default to encourage use, but allows you to get more information if you want it.  Perhaps a one-liner similar to...

* :information_source: Supported by ```${server name}```
* :information_source: Unknown or disconnected from ```${server name}```
* :warning: Possibly not supported by ```${server name}```

But with better phrasing, icons from system theme, and a ```More details...``` button that pops up a dialog with longer text akin to what you've written, maybe even a link to the IRCv3 page.

digitalcircuit commented:
I've implemented a simple label and icon with a ```Details...``` button, updated the pull request text and screenshots, and fixed a preexisting related bug, too (*SASL appeared enabled on first load when ```Use SASL Authentication``` was disabled*).

Let me know if you want any changes to the text.  I tried to come up with simple, good explanations, but I'm almost certain they can be improved.

digitalcircuit commented:
Quick update - fixed a regression with handling ```Use SASL Authentication``` enabling and disabling, and removed the link to the IRCv3 specifications as that's sort-of useless to those who aren't likely already familiar with IRC internals.  It'd be better to link to something like a wiki page on Quassel.

digitalcircuit commented:
Removed some mentions of authentication since SASL itself refers to authentication (*Simple Authentication and Security Layer*), and also makes the dialogs simpler to read.  To avoid resetting translations and maintain some hint of purpose, the ```Use SASL Authentication``` checkbox was *not* renamed.

Thanks to ```SaberUK``` and others on ```Freenode```/```#ircv3``` for the suggestion and discussion.
